src/Common/src/TypeSystem/Common/TypeSystemConstraintsHelpers.cs

```diff
@@ -97,6 +103,32 @@ private static bool CheckGenericSpecialConstraint(TypeDesc type, GenericConstrai
             return false;
         }
 
+        private static List<TypeDesc> GetInstantiatedConstraints(TypeDesc type, Instantiation typeInstantiation, Instantiation methodInstantiation)
+        {
+            var instantiatedConstraints = new List<TypeDesc>();
+
+            if (type.IsGenericParameter)
+            {
+                GenericParameterDesc genericParam = (GenericParameterDesc)type;
+
+                foreach (var constraint in genericParam.TypeConstraints)
+                    instantiatedConstraints.Add(constraint.InstantiateSignature(typeInstantiation, methodInstantiation));
+            }
```

Comment:
Do we need to do this recursively like https://github.com/dotnet/coreclr/blob/44f57065649af5f8bcbb7c71d827221a7bc1bf7a/src/vm/typedesc.cpp#L1990 ?

Reply:
I actually already implemented this recursively, but then reverted it after finding out that the following code generates a compiler error:
```
class Generic1<T, U> where T : U where U : class
{
    Generic2<T> gen2;
}

class Generic2<T> where T : class { }
```
But I just realized that it does indeed work with an Interface / concrete class like say `IComparable` instead of `class`. I assume this means that only special constraints are not resolved recursively.

Reply:
I came across a problem while trying out some more cases, which is not resolvable with the current information given to the `CheckConstraints` methods.
Before I re-implement this logic, do you know if there is anything similar to the `InstantiationContext` class of the CoreCLR already implemented in CoreRT?

Reply:
@MichalStrehovsky Do we have something similar to InstantiationContext  in CoreRT ?

Reply:
> something similar to InstantiationContext in CoreRT ?

We don't have it.

Reply:
I am not sure how to resolve this:
When calling `CheckConstraints` on `Generic2` of my previous example, the GenericParam `T` will have the uninstantiated constraint `!1`. In order to instantiate this constraint correctly I would need to know about the `Generic1` type inside `CheckConstraints`.
According to its comments the CoreCLR implemented the `InstantiationContext` class specifically for this kind of scenario.
PEVerify seems to somehow instantiate the constraints during the loading of the type itself.
I think this problem could also be resolved by simply adding a `Parent` property to the `GenericParameter`, which can then be used to instantiate the constraints. It would probably be best to instantiate the constraints at the same time as the type is instantiated itself, instead of having to instantiate them later, but I haven't figured out a good way to do this yet. 
Furthermore I noticed that the `CheckValidInstantiationArguments` method returns false for generic parameters (https://github.com/dotnet/corert/blob/bceb1952e884253ec88a7ec518116edf1cef9e0b/src/Common/src/TypeSystem/Common/TypeSystemConstraintsHelpers.cs#L100).
This explains why all these generic instantiation specific checks were not implemented yet. Shouldn't this method return true for generic parameters, or am I missing something here?

Reply:
> I think this problem could also be resolved by simply adding a `Parent` property to the `GenericParameter`

@MichalStrehovsky @davidwrighton Does this sound reasonable?

> `CheckValidInstantiationArguments` Shouldn't this method return true for generic parameters,

This method is used to validate the arguments of physical instantiations in the compiler/JIT (e.g. types that can be used to allocate objects, etc.). The physical instantiations cannot be over generic parameters.

Reply:
> @MichalStrehovsky @davidwrighton Does this sound reasonable?

Yes, that sounds reasonable. `NativeFormatGenericParameter` can simply implement it as a throwing property with a comment saying that we can implement it, but we would need to update the metadata schema to carry the parent information. We don't need that right now and not carrying that information is a size on disk saving optimization.